### ISSUE TITLE
`errors` property is missing in instance type of custom error

### ISSUE THREAD

**Issue by bhvngt** [ehmicky/modern-errors#15]
### Guidelines

- [X] Please search other issues to make sure this bug has not already been reported.
- [X] If this is related to a typo or the documentation being unclear, please click on the relevant page's `Edit` button (pencil icon) and suggest a correction instead.

### Describe the bug

I am exporting type declaration of my custom error with `InstanceType<typeof CustomError>`. Is this the right way to declare type for CustomErrors?

While declaring using this method I have noticed that the `errors` property to be missing on the exported type.

This can be reproduce with following code

```typescript
export const AppError = BaseError.subclass("AppError", {props: {prop1: undefined as string|undefined}})
export type AppErrorType = InstanceType<typeof AppError>
const err: AppErrorType = new AppError("custom error");
err.errors
```
I get following error when I do type checking.

```
 error TS2339: Property 'errors' does not exist on type 'SpecificErrorInstance<({ name: "serialize"; getOptions: (options: Options) => Options; isOptions: (options: unknown) => boolean; staticMethods: { serialize: (info: Omit<CommonInfo<never>, "error">, error: unknown) => ErrorObject; parse: (info: Omit<...>, errorObject: unknown) => ErrorInstance; }; instanceMethods: { ...'.
```

### Steps to reproduce

<img width="864" alt="image" src="https://user-images.githubusercontent.com/79074469/218724307-b7ccf49f-6c82-412b-869f-d9e87c334a55.png">

### Configuration

I am using webstorm as editor. I am using typescript `5.0.0-beta`

### Environment

System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 68.75 MB / 32.00 GB
    Shell: 3.6.0 - /opt/homebrew/bin/fish
  Binaries:
    Node: 18.14.0 - ~/Library/pnpm/node
    npm: 9.3.1 - ~/Library/pnpm/npm
  Browsers:
    Chrome: 109.0.5414.119
    Chrome Canary: 112.0.5594.3
    Firefox: 99.0
    Firefox Developer Edition: 107.0
    Safari: 16.2
  npmPackages:
    modern-errors: ^5.5.0 => 5.5.0

### Pull request (optional)

- [ ] I can submit a pull request.

**Comment by ehmicky** [comment]
Hi @bhvngt,

Thanks for reaching out again, it's great to get some feedback on the rougher points of the types.

Just to make sure, you're talking about the `errors` property documented [here](https://github.com/ehmicky/modern-errors#aggregate-errors), is that correct?

**Comment by bhvngt** [comment]
That true @ehmicky.

**Comment by ehmicky** [comment]
The `errors` property is only present when using the `errors` option.

```ts
new AppError("custom error", { errors: [childError] });
```

In other words, when the `errors` option is not passed, `error.errors` is not defined.

**Comment by bhvngt** [comment]
I am sorry, I gave you an incorrect example.

Here's the reproduction of my issue

```typescript
export const AppError = BaseError.subclass("AppError", {props: {prop1: undefined as string|undefined}})
export type AppErrorType = InstanceType<typeof AppError>

const aggErr1 = new AppError("custom error", {errors: [new Error("dsfdsfs")]});
aggErr1.errors // ✅ 

const aggErr2: AppErrorType = new AppError("custom error", {errors: [new Error("dsfdsfs")]});
aggErr2.errors // ❌
```

`aggErr2.errors` gives me following TS error - `TS2339: Property 'errors' does not exist on type 'SpecificErrorInstance({ name: "serialize"; getOptions: (options: Options) => Options; isOptions: (options: unknown) => boolean; staticMethods: { serialize: (info: Omit  , "error">, error: unknown) => ErrorObject; parse: (info: Omit...>, errorObject: unknown) => ErrorInstance; }; instanceMethods: { ...'.`

For now, I have a simple workaround. I declare AppErrorType as 
```typescript 
export type AppErrorType = InstanceType<typeof AppError> & {errors?: Error[]};
``` 

Would it make sense to declare `errors` as an optional property for custom error types?

**Comment by ehmicky** [comment]
Thanks for the additional information.

I think this would make sense indeed. Basically `AppErrorType["errors"]` should be `Error[] | undefined`, considering the `errors` property might exist, providing the `errors` option is passed.

Unfortunately, this is quite tricky to implement. The main files for those types are [here](https://github.com/ehmicky/modern-errors/blob/main/src/merge/cause.d.ts) and [there](https://github.com/ehmicky/modern-errors/blob/main/src/merge/aggregate.d.ts). What's challenging is to make all the [existing type tests](https://github.com/ehmicky/modern-errors/blob/main/src/merge/aggregate.test-d.ts) pass. They cover quite many edge cases, such as:
  - Passing non-`Error` instances to the `errors` option converts them to `Error` instances
  - Passing `as const` values to the `errors` option keeps them as is. For example, passing `errors: [error] as const` makes `error.errors` type `[Error]`, not `Error[]`
  - When passing both the `cause` and `errors` options, `cause.errors` is merge to `errors`, with lower priority. This is recursive, i.e. `cause.cause.errors` is also taken into account

This resulted in rather complex types. :( 

Would you be interested into giving it a shot and trying to fix the types, while still making the existing type tests work?

**Comment by bhvngt** [comment]
hmm.. sure. I will take a stab at it. I did glance through code and realised that its rather complex for my level of typescript knowledge. I guess if not anything, I will end up learning some of the typescript techniques that you have used here.

**Comment by ehmicky** [comment]
Yes, this is surprisingly complex! On the surface, it looks like it should be simple, but the types are actually guessing lots of things and the overall API is very dynamic and recursive. Types were actually the hardest part of this project, and now most of the issues so far seem to be (unsurprisingly) about them. On the plus side, I have added a very high amount of type tests.

If you do not have the time to look into it, do not worry, as I will most likely also try to find a solution. I am in a rather busy period at the personal level right now, which makes it hard to allocate a few hours focusing on this. However, I do intend to get this fixed. If you do want to get a stab at it though, thanks a lot for your help! :)

**Comment by bhvngt** [comment]
Completely understand @ehmicky. Thank you for this wonderful library. It has made my error management super clean. Guess all the effort behind the type management system is totally worth it as it works like a magic. 

Will give it a stab sometime this week.

**Comment by bhvngt** [comment]
I took a stab at this issue. Here's my short analysis

1. For `InstanceType<typeof CustomError>` to provide `errors` as optional property, [SpecificErrorInstance](https://github.com/ehmicky/modern-errors/blob/73ffaf647d2cfb72b036f808b2d5e2464b86cad3/src/merge/cause.d.ts#L17) should return `errors` as optional property.  This will require `AggregateErrorsProperty` to contain `errors` as an optional property.

2. Since [AggregateErrorsProperty](https://github.com/ehmicky/modern-errors/blob/73ffaf647d2cfb72b036f808b2d5e2464b86cad3/src/merge/aggregate.d.ts#L62) return `errors` only if it is either contained within `cause` or is passed in options during instance creation, any change here will break the related tests.

Thinking aloud, this library has been designed to infer `errors` during instance creation. Whereas, type inference of custom error happens from the `typeof` of Class . Maybe `InstanceType<typeof RepoError> & {errors?: Error[]};` would be the easiest and safest way to provide this property to the type inference.  Any alteration to accommodate this  might result in direct conflict with the dynamic inference of `errors`. With this approach, it might suffice to add this as part of the documentation.

Would love to hear your thoughts.

**Comment by ehmicky** [comment]
Hi @bhvngt,

Sorry for the late reply!

## Core problem

I think your analysis is right on point. Especially, the problem is the conflict between the two things you are outlining:
  1. `InstanceType<typeof ErrorClass>["errors"]` should be defined and be an optional property. Otherwise, one would run in the problem described in this issue.
  2. `new ErrorClass().errors` should fail since `errors` would not exist at runtime.

## Solving that problem

However, doing both is difficult. The core issue is related to using generic parameters on constructors. Right now, the error class's generic parameters are:

https://github.com/ehmicky/modern-errors/blob/5f2d7b48390be4c9dac271512dc90b2d1369104d/src/subclass/create.d.ts#L61-L65

When instantiated as `new ErrorClass({ errors })`, `AggregateErrorsArg` is inferred based on the `errors` option. In most cases, `error.errors`'s type will be the same as the `errors` option.

https://github.com/ehmicky/modern-errors/blob/5f2d7b48390be4c9dac271512dc90b2d1369104d/src/options/instance.d.ts#L44

When instantiated without any `errors` option, i.e. as `new ErrorClass()`, the default type `= AggregateErrors` is used. 
When using `InstanceType<typeof ErrorClass>` is used, `extends AggregateErrors` is used instead.
In both cases, that `AggregateErrors` type eventually results in `error.errors` not being defined.

## Changing the current type

One possible solution might be to declare the generic types like this instead:

```ts
   AggregateErrorsArg extends AggregateErrors = undefined, 
```

This would allow us to distinguish between `new ErrorClass()` (which would use `= undefined`) and `InstanceType<typeof ErrorClass>` (which would use `extends AggregateErrors`).

However, when extending a `class`, TypeScript uses the default types of generic parameters. For example:

```ts
const ExampleError = ModernError.subclass('ExampleError', {
  custom: class extends ModernError {},
})
```

In that instance, `extends ModernError` would use `= undefined`, i.e. `class extends ModernError {}` requires `errors` to be `undefined`. However, this conflicts with the `custom` option, which should allow any `errors`, i.e. be typed as `AggregateErrors` instead.

## Coming back to the initial situation

Based on the above, the only way I manage to make inheritance work is to use the same generic ctype for `extends ...` and `= ...`. In other words, using:

```ts
   AggregateErrorsArg extends AggregateErrors = AggregateErrors, 
```

However, this has the following consequence: `InstanceType<typeof CustomError>["errors"]` and `new ErrorClass().errors` will have the same type. At the moment, I think that runtime type inference might be more important, which leads to your initial problem.

---

As you can see, this is quite a hard typing problem! I am happy to merge any PR that solves this, while still keeping the current type tests as valid.

In the meantime, a type assertion can work around this problem:

```ts
const aggErr2: AppErrorType = new AppError("custom error", {errors: [new Error("dsfdsfs")]});
(aggErr2 as { errors?: Error[] }).errors
```

What do you think?

**Comment by bhvngt** [comment]
I agree that creating either type assertion or explicitly merging `errors` property to `InstanceType<typeof CustomError>` is more favourable then changing the core design of type system. 

Issue occurs when Type creation is required for those Custom Error type who are expected to aggregate or wrap errors. I guess explicitly merging `errors` property for those types is a simple enough trade off. Otherwise, we will need to remove runtime property inference in favour of hard coding `errors` for all custom errors. This may end up making all custom errors as kind of aggregated errors. This may not always be the case. 

Maybe a simple note in the documentation may help.

**Comment by ehmicky** [comment]
I think I found a solution, although not perfect. I just released it in [`5.5.3`](https://github.com/ehmicky/modern-errors/releases/tag/5.5.3), and the behavior is:
  - `new ErrorClass({ errors }).errors` gets the right type, i.e. same as the `errors` being passed as option
  - `new ErrorClass().errors` is typed as `Error[] | undefined`. In principle, it should be `undefined` or not being defined at all, so this is imperfect, but that's a tradeoff :-/
  - `InstanceType<typeof ErrorClass>["errors"]` is typed as `Error[] | undefined`. This is your original case, so I think this should fix your problem.

Please let me know what you think!

**Comment by bhvngt** [comment]
This is great. This will certainly save me some typing. I use [neverthrow](https://github.com/supermacro/neverthrow), so I like to return errors instead of throwing them. That allows me to handle errors more explicitly with additional benefit of type safety. Many of my application errors are wrapper around library errors. I would have had to use merge error property in all most all cases. 

> new ErrorClass().errors is typed as Error[] | undefined. In principle, it should be undefined or not being defined at all, so this is imperfect, but that's a tradeoff :-/

I guess this tradeoff though not as perfect as before, communicates well when anyone is trying to access `undefined` errors. So, it looks good to me.

Thanks a lot for your time, patience and wonderful solution. I will close this thread, given that you have already released the solution.
